src/core/params.hpp

```diff
@@ -0,0 +1,138 @@
+      const IterableParam *params = getIterableParams(numParams);               \
+      arrayResize(result, numParams);                                           \
+      for (size_t i = 0; i < numParams; i++) {                                  \
+        result.elements[i] = *const_cast<ParameterInfo *>(params[i].paramInfo); \
+      }                                                                         \
+      return result;                                                            \
+    }();                                                                        \
+    return result;                                                              \
+  }
+
+// Implements setParam()/getParam()
+#define PARAM_GET_SET()                                             \
+  void setParam(int index, const SHVar &value) {                    \
+    size_t numParams;                                               \
+    const IterableParam *params = getIterableParams(numParams);     \
+    if (index < numParams) {                                        \
+      params[index].setParam(params[index].getVarPtr(this), value); \
+    } else {                                                        \
+      throw InvalidParameterIndex();                                \
+    }                                                               \
+  }                                                                 \
+  SHVar getParam(int index) {                                       \
+    size_t numParams;                                               \
+    const IterableParam *params = getIterableParams(numParams);     \
+    if (index < numParams) {                                        \
+      return params[index].getParam(params[index].getVarPtr(this)); \
+    } else {                                                        \
+      throw InvalidParameterIndex();                                \
+    }                                                               \
+  }
+
+// Implements warmup(ctx) for parameters
+// call from warmup manually with context
+#define PARAM_WARMUP(_ctx)                                      \
+  {                                                             \
+    size_t numParams;                                           \
+    const IterableParam *params = getIterableParams(numParams); \
+    for (size_t i = 0; i < numParams; i++)                      \
+      params[i].warmup(params[i].getVarPtr(this), _ctx);        \
+  }
+
+// implements cleanup() for parameters
+// call from cleanup manually
+#define PARAM_CLEANUP()                                         \
+  {                                                             \
+    size_t numParams;                                           \
+    const IterableParam *params = getIterableParams(numParams); \
+    for (size_t i = 0; i < numParams; i++)                      \
+      params[i].cleanup(params[i].getVarPtr(this));             \
```

Comment:
`cleanup` should probably use the reverse order. Not sure it is really needed but it's kind of a convention to "destroy" in reverse "creation" order.

Reply:
correct, cleanup is safer in reverse order definitely

Reply:
Remind me to fix ShardsVar in rust like that too :)

Reply:
Can't imagine a case where it matters for params but I can add it for consistency's sake

Reply:
True but could be